rtos/TARGET_CORTEX/TOOLCHAIN_GCC_ARM/mbed_boot_gcc_arm.c

```diff
@@ -0,0 +1,125 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cmsis.h"
+#include "mbed_toolchain.h"
+#include "mbed_boot.h"
+
+#if defined(MBED_CONF_RTOS_PRESENT)
+#include "mbed_rtos_storage.h"
+#include "cmsis_os2.h"
+
+static osMutexId_t               malloc_mutex_id;
+static mbed_rtos_storage_mutex_t malloc_mutex_obj;
+static osMutexAttr_t             malloc_mutex_attr;
+
+static osMutexId_t               env_mutex_id;
+static mbed_rtos_storage_mutex_t env_mutex_obj;
+static osMutexAttr_t             env_mutex_attr;
+#endif
+
+/* Define stack sizes if they haven't been set already */
+#if !defined(ISR_STACK_SIZE)
+    #define ISR_STACK_SIZE ((uint32_t)1024)
+#endif
+
+/* Defined by linker script */
+extern uint32_t __end__[];
+
+extern void __libc_init_array (void);
+
+/*
+ * mbed entry point for the GCC toolchain
+ *
+ * Override gcc boot hook software_init_hook to run code before main.
+ */
+void software_init_hook(void)
+{
+    unsigned char *free_start = (unsigned char*)__end__;
+    uint32_t free_size = (uint32_t)INITIAL_SP - (uint32_t)__end__;
+
+#ifdef ISR_STACK_START
+    /* Interrupt stack explicitly specified */
+    mbed_stack_isr_size = ISR_STACK_SIZE;
+    mbed_stack_isr_start = ISR_STACK_START;
+#else
+    /* Interrupt stack -  reserve space at the end of the free block */
+    mbed_stack_isr_size = ISR_STACK_SIZE < free_size ? ISR_STACK_SIZE : free_size;
+    mbed_stack_isr_start = free_start + free_size - mbed_stack_isr_size;
+    free_size -= mbed_stack_isr_size;
+#endif
+
+    /* Heap - everything else */
+    mbed_heap_size = free_size;
+    mbed_heap_start = free_start;
+
+
+    mbed_init();
+    mbed_rtos_start();
+}
+
+void mbed_toolchain_init()
+{
+#if defined(MBED_CONF_RTOS_PRESENT)
```

Comment:
I don't think we need this checks, nothing in `rtos/` folder will be compiled in if the RTOS is disabled.